### PR TITLE
rx-ified the ReservationReportCachingAgent

### PR DESCRIPTION
**Pull request by ajordens**
- Each account will be loaded in parallel

### CODE REVIEW

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy [hunk 1]
```diff
@@ -110,90 +117,105 @@ class ReservationReportCachingAgent implements CachingAgent {
 
   @Override
   CacheResult loadData(ProviderCache providerCache) {
+    long startTime = System.currentTimeMillis()
     log.info("Describing items in ${agentType}")
-    Map<String, AmazonReservationReport.OverallReservationDetail> reservations = [:].withDefault { String key ->
+
+    Map<String, AmazonReservationReport.OverallReservationDetail> reservations = new ConcurrentHashMap<>()
+    def getReservation = { String key ->
```
Comment (ajordens) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy]: Helper method that wraps the ConcurrentHashMap accessing in a way not too dissimilar to [:].withDefault()

**Comment by ajordens** [comment]
PTAL @cfieber

**Comment by ajordens** [comment]
Locally this change looks 20-30%+ than previous.

I'm going to look at edda vs. aws for the describeReservedInstances() since we're already going to edda for describeInstances().

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy [hunk 2]
```diff
@@ -110,90 +117,105 @@ class ReservationReportCachingAgent implements CachingAgent {
...
-              def osTypeName = operatingSystemType(it.platform ? "Windows" : "Linux/UNIX").name
-              def reservation = reservations["${it.placement.availabilityZone}:${osTypeName}:${it.instanceType}"]
-              reservation.totalUsed.incrementAndGet()
 
-              if (it.vpcId) {
-                reservation.accounts[credentials.name].usedVpc.incrementAndGet()
-              } else {
-                reservation.accounts[credentials.name].used.incrementAndGet()
+          def describeInstancesRequest = new DescribeInstancesRequest()
+          def allowedStates = ["pending", "running", "shutting-down", "stopping", "stopped"] as Set<String>
```
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy]: could pull this up somewhere outside the loop

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy (hunk 1 above)
Comment (cfieber) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy]: both places this is called it is assembling the key out of the component parts to call it .. could parameterize the closure with the parts, and construct the key inside here

**Comment by cfieber** [comment]
LGTM

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy (hunk 1 above)
Comment (ajordens) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy]: Done

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy (hunk 2 above)
Comment (ajordens) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/ReservationReportCachingAgent.groovy]: Didn't really seem to add much value pulling it out of the loop so I've opt'd to leave it.

**Pull request merged**